node/stateSync.go

```diff
@@ -0,0 +1,201 @@
+			time.Sleep(3 * time.Second)
+			select {
+			case <-s.done:
+				return
+			default:
+			}
+		}
+	}()
+
+	for s.sched.Pending() > 0 {
+		if err = s.commit(false); err != nil {
+			return err
+		}
+
+		for i := 0; i < len(s.peers); i++ {
+			jobs, _ := s.sched.Missing(config.MaxStateReqSize)
+			if len(jobs) == 0 {
+				break
+			}
+			s.hashesCh <- jobs
+		}
+		select {
+		case data := <-s.stateCh:
+			for _, blob := range data {
+				_, err := s.processNodeData(blob)
+				switch err {
+				case nil:
+					s.bytesUncommitted += len(blob)
+				default:
+					log.Info("processNodeData err:", err)
+				}
+			}
+		case <-s.done:
+			return
+		}
+	}
+	return err
+}
+
+func (s *stateSync) commit(force bool) error {
+	if !force && s.bytesUncommitted < config.IdealBatchSize {
+		return nil
+	}
+	s.d.database.NewBatch()
+	if err := s.sched.Commit(s.d.database); err != nil {
+		return err
+	}
+	if err := s.d.database.BatchCommit(); err != nil {
+		return err
+	}
+	s.bytesUncommitted = 0
+	return nil
+}
+
+// processNodeData tries to inject a trie node data blob delivered from a remote
+// peer into the state trie, returning whether anything useful was written or any
+// error occurred.
+func (s *stateSync) processNodeData(blob []byte) (common.Uint256, error) {
+	res := trie.SyncResult{Data: blob}
+	var err error
+	res.Hash, err = common.Uint256ParseFromBytes(s.hasher(blob))
+	if err != nil {
+		return common.EmptyUint256, err
+	}
+	err = s.sched.Process(res)
+	return res.Hash, err
+}
+
+func (s *stateSync) startWorker(ctx context.Context, workerId uint32, p *RemoteNode) {
+	var hashes []common.Uint256
+	for {
+		select {
+		case <-s.stopChans[workerId]:
+			return
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		select {
+		case hashes = <-s.hashesCh:
+		case hashes = <-s.failedHashesCh:
+		case <-s.stopChans[workerId]:
+			return
+		case <-ctx.Done():
+			return
+		}
+
+		resp, err := p.GetStates(hashes)
+
+		if err != nil {
+			s.failedHashesCh <- hashes
+			continue
+		}
+		if len(resp) > 0 {
+			s.stateCh <- resp
+		}
+	}
+}
+
+// run starts the task assignment and response processing loop, blocking until
+// it finishes, and finally notifying any goroutines waiting for the loop to
+// finish.
+func (s *stateSync) run() {
+	s.err = s.loop()
+	close(s.done)
+}
+
+// newStateSync creates a new state trie download scheduler. This method does not
+// yet start the sync. The user needs to call run to initiate.
+func newStateSync(d *Downloader, root common.Uint256, peers []*RemoteNode) *stateSync {
+	return &stateSync{
+		d:              d,
+		root:           root,
+		sched:          trie.NewSync(root),
+		hasher:         trie.Sha256Key,
+		done:           make(chan struct{}),
+		stateCh:        make(chan [][]byte),
+		stopChans:      make([]chan struct{}, len(peers)),
+		hashesCh:       make(chan []common.Uint256, config.Parameters.SyncStatesBatchSize),
+		failedHashesCh: make(chan []common.Uint256, config.Parameters.SyncStatesBatchSize),
```

Comment:
This channel is not needed at all. When fails, put it back to the priority queue